objects/host_checkstop_obj.c

```diff
@@ -0,0 +1,165 @@
+#include "interfaces/openbmc_intf.h"
+#include "openbmc.h"
+#include "gpio.h"
+
+static const gchar* dbus_object_path = "/org/openbmc/control";
+static const gchar* instance_name = "checkstop0";
+static const gchar* dbus_name = "org.openbmc.control.Checkstop";
+
+static GDBusObjectManagerServer *manager = NULL;
+
+GPIO checkstop = (GPIO){ "CHECKSTOP" };
+
+static gboolean
+chassis_reboot(gpointer connection)
+{
+    int rc = 0;
+    uint8_t gpio = 0;
+    GDBusProxy *proxy;
+    GError *error;
+    GVariant *parm = NULL;
+    GVariant *result = NULL;
+
+    // The gpio line may flicker during power on/off, so check that the value
+    // is still 0 (checkstopped) in order to reboot
+    rc = gpio_open(&checkstop);
+    if (rc != GPIO_OK) {
+        return FALSE;
+    }
+    rc = gpio_read(&checkstop, &gpio);
+    if (rc != GPIO_OK) {
+        return FALSE;
+    }
+    gpio_close(&checkstop);
+    if (!gpio)
+    {
+        printf("Host Checkstop, rebooting host\n");
+        error = NULL;
+        proxy = g_dbus_proxy_new_sync((GDBusConnection*)connection,
```

Comment:
Sounds like a plan.

The output of the call is a dictionary with unique connection name as the keys, and interface list as the values.  Its ok to hardcode the interface you want to use.

As for which service - I implemented the mapper ( for better or for worse ) such that multiple services can implement the same object path.  So a generic way could be that you search for your desired interface in each of the service->[interfaces] in the dictionary.

It would probably make sense to reverse this in the mapper such that the keys are the interfaces and the values are the unique-service name.  Then you could just do:

info = mapper.GetObject('/org/openbmc/foo')
service = info['the.interface.i.want']

I'll add this to the issue.